src/main/kotlin/DirectoryLogic.kt

```diff
@@ -2,20 +2,57 @@ import java.lang.IllegalArgumentException
 
 class DirectoryLogic {
 
-    private val supportedExtensions = arrayOf("wav", "mp3", "jpeg", "tr")
+    private val supportedExtensions = arrayOf("wav", "mp3", "jpeg", "jpg", "tr")
+    private val supportedContainers = arrayOf("tr")
+    private val compressedTypes = arrayOf("mp3", "jpeg", "jpg")
+    private val groupings = arrayOf("book", "chapter", "verse", "chunk")
 
+    @Throws(IllegalArgumentException::class)
     fun buildFullFilePath(
-        inputFilePath: String,
-        languageCode: String,
-        dublinCoreId: String,
-        group: String,
-        projectId: String = "",
-        mediaExtension: String = "",
-        mediaQuality: String = "hi"
+            inputFilePath: String,
+            languageCode: String,
+            dublinCoreId: String,
+            group: String,
+            projectId: String = "",
+            mediaExtension: String = "",
+            mediaQuality: String = ""
     ): String {
 
-        return ""
+        val fileExt = inputFilePath.split("/").last().split('.').last()
+
+        var path = ""
+
+        if (languageCode.isBlank()) throw IllegalArgumentException("Language code is empty")
+        path += "$languageCode/"
+
+        if (dublinCoreId.isBlank()) throw IllegalArgumentException("Dublin Code ID is empty")
+        path += "$dublinCoreId/"
+
+        if (group.isBlank()) throw IllegalArgumentException("Group is empty")
+        if (!groupings.contains(group)) throw IllegalArgumentException("Group is not supported")
+
+        if (projectId.isNotBlank()) path += "$projectId/"
+
+        path += "CONTENTS/"
+        path += "$fileExt/"
+
+        if (supportedContainers.contains(fileExt)) {
+            if (mediaExtension.isBlank()) throw IllegalArgumentException("Media Extension is empty")
+            if (!supportedExtensions.contains(mediaExtension)) throw IllegalArgumentException("Media Extension is not supported")
+            path += "$mediaExtension/"
+        } else if (!supportedExtensions.contains(fileExt)) throw IllegalArgumentException(".$fileExt file is not supported")
+
+        // if the file is a compressed type
+        if (compressedTypes.contains(fileExt) || compressedTypes.contains(mediaExtension)) {
+            path += if (mediaQuality.isBlank()) "hi/"
```

Comment:
"hi" and "low" are constants at the very least, and should get defined above (makes it easier to reassign them if these values change without having to go hunting through the code)

if this was an enum type you could do:
```
"$MediaQuality.of(mediaQuality)/"
```
which would handle knowing if it's hi or low, or throw an exception.
If those qualities change, or more are added, the code here wouldn't need to change.

At the very least you could put your quality options in an array and be consistent with how you handle the other constant types and do a .contains() call below.